File: Problems/CourseScheduleII.java
/**
 * There are a total of numCourses courses you have to take, labeled from 0 to numCourses - 1. 
 * You are given an array prerequisites where prerequisites[i] = [ai, bi] indicates that you must take course bi first if you want to take course ai.
 * 
 * For example, the pair [0, 1], indicates that to take course 0 you have to first take course 1.
 * Return the ordering of courses you should take to finish all courses. If there are many valid answers, 
 * return any of them. If it is impossible to finish all courses, return an empty array.
 */
package Problems;

import java.util.*;

public class CourseScheduleII {
    public static int[] findOrder(int numCourses, int[][] prerequisites) {
        if (numCourses == 0)
            return null;
        int[] indegree = new int[numCourses];
        int[] res = new int[numCourses];
        int index = 0;
        for (int i = 0; i < prerequisites.length; i++)
            indegree[prerequisites[i][0]]++;

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++)
            if (indegree[i] == 0) {
                res[index++] = i;
                queue.offer(i);
            }

        while (!queue.isEmpty()) {
            int prerequisite = queue.poll();
            for (int i = 0; i < prerequisites.length; i++) {
                if (prerequisites[i][1] == prerequisite) {
                    indegree[prerequisites[i][0]]--;
                    if (indegree[prerequisites[i][0]] == 0) {
                        res[index++] = prerequisites[i][0];
                        queue.offer(prerequisites[i][0]);
                    }
                }
            }
        }

        return (index == numCourses) ? res : new int[0];
    }

    public static void main(String[] args) {
        int numCourses = 3;
        int[][] prerequisites = { { 0, 1 }, { 0, 2 }, { 1, 2 } };
        int[] res = findOrder(numCourses, prerequisites);
        for (int i : res) {
            System.out.print(i + " ");
        }
    }
}

File: Problems/CourseScheduleII.cpp
#include <iostream>
#include <queue>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
   public:  // solution using Topological sorting
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> inDegree(numCourses, 0);
        unordered_map<int, vector<int>> outDegree;
        for (auto& courses : prerequisites) {
            inDegree[courses[0]]++;
            outDegree[courses[1]].push_back(courses[0]);
        }
        queue<int> q;
        for (int i = 0; i < inDegree.size(); i++) {
            if (inDegree[i] == 0) {
                q.push(i);
            }
        }
        vector<int> res;
        while (!q.empty()) {
            int course = q.front();
            q.pop();
            res.push_back(course);
            vector<int> v = outDegree[course];
            for (int i : v) {
                inDegree[i]--;
                if (inDegree[i] == 0)
                    q.push(i);
            }
        }
        return res.size() == numCourses ? res : vector<int>();
    }
};